# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/base/SelectionManager.cpp
# Commit: 69d2ec7b4d6c
# Full Hash: 69d2ec7b4d6ca35da23f14602cdf78d939da40cb
# Author: James Teh <jteh@mozilla.com>
# Date: 2022-05-27 15:58:57
# Regressor Bug: 1737919
# File Overlap Count: 1
# Description:
#   Bug 1737919 part 4: Update cached spelling errors when a spelling error is added without the text changing. r=morgan,smaug
#   
#   We already have an nsISelectionListener, but that only tells us that a change happened somewhere in the selection, not which range changed.
#   We don't want to push a cache update for all ranges when only one changed.
#   Therefore, this patch adds an accessibility notification in mozInlineSpellChecker::AddRange.
# ==============================================================================

diff -r 388a15605114 -r 69d2ec7b4d6c accessible/base/SelectionManager.cpp
--- a/accessible/base/SelectionManager.cpp	Fri May 27 10:56:41 2022 +0000
+++ b/accessible/base/SelectionManager.cpp	Fri May 27 10:56:41 2022 +0000
@@ -13,11 +13,13 @@
 #include "nsCoreUtils.h"
 #include "nsEventShell.h"
 #include "nsFrameSelection.h"
+#include "TextLeafRange.h"
 
 #include "mozilla/PresShell.h"
 #include "mozilla/dom/Document.h"
 #include "mozilla/dom/Selection.h"
 #include "mozilla/dom/Element.h"
+#include "mozilla/StaticPrefs_accessibility.h"
 
 using namespace mozilla;
 using namespace mozilla::a11y;
@@ -226,4 +228,14 @@
   }
 }
 
+void SelectionManager::SpellCheckRangeAdded(const nsRange& aRange) {
+  // Events are fired in SelectionManager::NotifySelectionChanged. This is only
+  // used to push cache updates.
+  if (StaticPrefs::accessibility_cache_enabled_AtStartup()) {
+    dom::Document* doc = aRange.GetStartContainer()->OwnerDoc();
+    MOZ_ASSERT(doc);
+    TextLeafPoint::UpdateCachedSpellingError(doc, aRange);
+  }
+}
+
 SelectionManager::~SelectionManager() = default;